plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java

```diff
@@ -496,6 +501,172 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
         }
     }
 
+    @Test
+    public void testJoinPushdownWithImplicitCast()
+    {
+        try (TestTable leftTable = new TestTable(getQueryRunner()::execute, "left_table", "(id int, varchar_50 varchar(50))", ImmutableList.of("(1, 'India')", "(2, 'Poland')"));
+                TestTable rightTable = new TestTable(getQueryRunner()::execute, "right_table_", "(varchar_100 varchar(100), varchar_unbounded varchar)", ImmutableList.of("('India', 'Japan')", "('France', 'Poland')"))) {
+            String leftTableName = leftTable.getName();
+            String rightTableName = rightTable.getName();
+            Session session = joinPushdownEnabled(getSession());
+
+            // Implict cast between bounded varchar
+            assertJoinFullyPushedDown(session, leftTableName, rightTableName, "LEFT JOIN", "l.varchar_50 = r.varchar_100", asList(1, 2));
+            assertJoinFullyPushedDown(session, leftTableName, rightTableName, "RIGHT JOIN", "l.varchar_50 = r.varchar_100", asList(1, null));
+            assertJoinFullyPushedDown(session, leftTableName, rightTableName, "INNER JOIN", "l.varchar_50 = r.varchar_100", List.of(1));
+
+            // Implict cast between bounded and unbounded varchar
+            assertJoinFullyPushedDown(session, leftTableName, rightTableName, "LEFT JOIN", "l.varchar_50 = r.varchar_unbounded", asList(1, 2));
+            assertJoinFullyPushedDown(session, leftTableName, rightTableName, "RIGHT JOIN", "l.varchar_50 = r.varchar_unbounded", asList(null, 2));
+            assertJoinFullyPushedDown(session, leftTableName, rightTableName, "INNER JOIN", "l.varchar_50 = r.varchar_unbounded", List.of(2));
+        }
+    }
+
+    @Nested
+    class TestCastPushdown
+    {
+        @Test
+        public void testCastProjectionPushdown()
+        {
+            for (CastTestCase testCase : supportedCastTypePushdown()) {
+                try (TestTable table = new TestTable(getQueryRunner()::execute, "test_cast_", "(a %s)".formatted(testCase.fromType()))) {
+                    assertThat(query("SELECT CAST(a AS %s) FROM %s".formatted(testCase.castType(), table.getName())))
+                            .isFullyPushedDown();
+                }
+            }
+
+            for (CastTestCase testCase : unsupportedCastTypePushdown()) {
```

Comment:
ProjectionPushdown may be more restrictive than Join criteria pushdown. Please test for joins separately.